### PR TITLE
feat: Allow arbitrary destination address for token creation

### LINKED ISSUE #39: Token creation destination address should not be hardcoded
At the moment of writing these lines, only the Token Authority can create new tokens. 

Gwen hardcodes the destination address, i.e., the destination where the new tokens will be allocated, to the address of the user creating the token, which is always the token authority.

It would be great if the address were not hardcoded to simplify the Alpha Cohort onboarding documents.

### PR DESCRIPTION
**Pull request by stanleyjones**
I was getting started on #63 and I needed to change this feature anyway. Closes #39.

### BEFORE

Screenshot 2023-08-10 at 10 10 33 AM

### NOW

Screenshot 2023-08-10 at 10 11 15 AM

### WITH VALIDATION

Screenshot 2023-08-10 at 10 05 42 AM

### CODE REVIEW

**Comment by fmorency** [comment]
Creating a new token using the pre-populated "Destination Address" results in an error, even if the address is correctly formatted.



Also, the request made to the backend is malformed, resulting in a `500: Internal Error`. The issue resides in how the `Timestamp` is encoded in CBOR.

GOOD (many-rs):
```
18([
    h'a3012604581d0119f0ec2aca1eae0d7405d2bb69edb256be44c68378052193ddce93a4666b6579736574587181a7010202581d0119f0ec2aca1eae0d7405d2bb69edb256be44c68378052193ddce93a4032604810220012158206112eec037fb20b6c8e11a22ffa0b4842892d03986cbe29a45cb50cfda6da35622582082454c56fb76249416821e339dfcccda743a7f47a84027fc85e6c0087b8bbbd1',
    {},
    h'd92711a501d92710581d0119f0ec2aca1eae0d7405d2bb69edb256be44c68378052193ddce93a4036d746f6b656e732e637265617465045833a200a30064544553540163545354020901d92710581d0119f0ec2aca1eae0d7405d2bb69edb256be44c68378052193ddce93a405c11a64d65345075051ad338c07d4683eb1c56a794a26f9fb',
    h'cbe870676165a4e7e781c1d88f24e4f47e73901ad45236292e9b287f25db2cfee3cbc9068111169897dd17463c9f4ac9e3ffa89b97c7d33d2635387612a9a45c',
])
```

BAD (this PR):
```
18([
    h'a3012704581d01f2e272e51c5fd0b02abeeb8bbef05fee26e112eba264734919b867c9666b6579736574585081a6010102581d01f2e272e51c5fd0b02abeeb8bbef05fee26e112eba264734919b867c90327048102200621d8405820d1931c94ba5b226df8c5e2ebb71b0cdd6bd3f75d3bf7184a1c3fe31ed25a440a',
    {},
    h'd92711a600010178326d61687a6f6534786664727035626d626b78337679787078716c3778636e796973356f72676934326a64673467707369646a036d746f6b656e732e637265617465045874a300a30066466f6f6261720163464252020901d92710581d01f2e272e51c5fd0b02abeeb8bbef05fee26e112eba264734919b867c902a178326d61687a6f6534786664727035626d626b78337679787078716c3778636e796973356f72676934326a64673467707369646ac247038d7ea4c6800005c11a64d652550753d84050169ddceefe4b36aa00ec7435e4b41e4f',
    64(h'c0f9abfe60fde1aea676948f5707fae0e4e3b85f65d6b0210ad5fc3c98dd7f05390243a8eba92d50eb44b8adf3a16e9baf8542dc0e624dbbd9ca78e9f5430a05'),
])
```

Notice that the last field is tagged with a `64`, which is wrong.

**Comment by stanleyjones** [comment]
The `64` tag is for a Uint8Array and is probably because of the Buffer polyfill we were using. Now that we've rolled back to an earlier version I'm going to see if this works again.

**Comment by stanleyjones** [comment]
The request is now accepted by the backend (no internal server error) but the tokens are still being assigned to the sender and not the address specified. Investigating.

**Comment by stanleyjones** [comment]
Okay, I found and fixed a bug and the request seems to be correct now:

```
18([
     h'A3012704581D014A101D521D810211A0C6346BA89BD1CC1F821C03B969FF9D5C8B2F59666B6579736574584E81A6010102581D014A101D521D810211A0C6346BA89BD1CC1F821C03B969FF9D5C8B2F5903270481022006215820D59E31A5CD12E054BAF4CF8625AA6D810C93B91C0B28132FC09A8F3785CA45C6',
     {},
     h'D92711A400010178326D6166666261686B736477617165656E6179793267786B65333268676237617134616F3477743734356C7366733677696A70036B6C65646765722E696E666F05C11A64F0C821',
     h'6039815DB20A53207C407C82ED1481DAF2B09C2563058187052A45F905C0B54B377F699D45B0F0FC4819FD020603CCF2918670B30E58FE3DC653A78B85E86D08'
])
```

But unfortunately I get this error:

```
Error: Unauthorized to do this operation.
```

I'm sending as ID1 and can create tokens owned by ID1. I only get this when I try to create tokens owned by ID2. Any thoughts, @fmorency ?

**Comment by fmorency** [comment]
> Okay, I found and fixed a bug and the request seems to be correct now:
> 
> ```
> 18([
>      h'A3012704581D014A101D521D810211A0C6346BA89BD1CC1F821C03B969FF9D5C8B2F59666B6579736574584E81A6010102581D014A101D521D810211A0C6346BA89BD1CC1F821C03B969FF9D5C8B2F5903270481022006215820D59E31A5CD12E054BAF4CF8625AA6D810C93B91C0B28132FC09A8F3785CA45C6',
>      {},
>      h'D92711A400010178326D6166666261686B736477617165656E6179793267786B65333268676237617134616F3477743734356C7366733677696A70036B6C65646765722E696E666F05C11A64F0C821',
>      h'6039815DB20A53207C407C82ED1481DAF2B09C2563058187052A45F905C0B54B377F699D45B0F0FC4819FD020603CCF2918670B30E58FE3DC653A78B85E86D08'
> ])
> ```
> 
> But unfortunately I get this error:
> 
> ```
> Error: Unauthorized to do this operation.
> ```
> 
> I'm sending as ID1 and can create tokens owned by ID1. I only get this when I try to create tokens owned by ID2. Any thoughts, @fmorency ?

This is expected for a network without the "Token Create Migration" active.

Local dev environments need to enable the "Token Create Migration".

I would suggest you enable
1. Account Count Data Attribute
2. Memo Migration
3. Token Migration
4. Token Create Migration

when starting your network. Simply remove the `disabled` lines for those entries.

You will need to comment the `hash` entry of the staging file if you're enabling those.

**Comment by stanleyjones** [comment]
Okay, this is working now. I was assigning both the `owner` and the initial distribution address to the input address and it looks like the owner has to remain the sender.

File: src/api/services/queries/ledger.ts [hunk 1]
```diff
@@ -29,10 +29,10 @@ export function useTokenInfo(neighborhood: Network | undefined) {
...
     queries: tokenList
       ? [...tokenList.symbols.entries()].map(([address]) => ({
-          queryKey: [neighborhood?.url, "tokens.info", address],
-          queryFn: async () => await neighborhood?.tokens.info({ address }),
-          enabled: !!neighborhood,
-        }))
+        queryKey: [neighborhood?.url, "tokens.info", address],
+        queryFn: async () => await neighborhood?.tokens.info({ address }),
+        enabled: !!neighborhood,
+      }))
```
Comment (stanleyjones) [inline src/api/services/queries/ledger.ts:35]: Just auto-formatting. No code changes.

File: src/api/services/queries/ledger.ts [hunk 2]
```diff
@@ -42,27 +42,29 @@ export function useCreateToken(neighborhood: Network | undefined) {
...
+      symbol,
     }: {
-      address: string;
       amount: string;
+      destination: string;
       name: string;
-      symbol: string;
+      owner: string;
       precision?: number;
+      symbol: string;
```
Comment (stanleyjones) [inline src/api/services/queries/ledger.ts:57]: I sorted the arg list so the diff is a little messy, but I'm _removing_ `address` and replacing it with both `owner` and `destination`. The former will be the same as the sender and the latter will take the input address.

File: src/api/services/queries/ledger.ts [hunk 3]
```diff
@@ -42,27 +42,29 @@ export function useCreateToken(neighborhood: Network | undefined) {
...
     }) =>
       await neighborhood?.tokens.create({
         summary: {
           name,
           symbol: symbol.toUpperCase(),
-          precision: 9,
+          precision,
```
Comment (stanleyjones) [inline src/api/services/queries/ledger.ts:63]: `precision` already defaulted to `9` so we don't need to hardcode a value here.

File: src/api/services/queries/ledger.ts [hunk 4]
```diff
@@ -42,27 +42,29 @@ export function useCreateToken(neighborhood: Network | undefined) {
...
         },
-        owner: address,
+        owner,
```
Comment (stanleyjones) [inline src/api/services/queries/ledger.ts:65]: The `owner` parameter will now be set to the `owner` argument above...

File: src/api/services/queries/ledger.ts [hunk 5]
```diff
@@ -42,27 +42,29 @@ export function useCreateToken(neighborhood: Network | undefined) {
...
         distribution: {
-          [address]: BigInt(parseInt(amount) * 10 ** precision),
+          [destination]: BigInt(parseInt(amount) * 10 ** precision),
```
Comment (stanleyjones) [inline src/api/services/queries/ledger.ts:67]: ...while the destination of the initial distribution will be set to the `destination` argument.

File: src/pages/services/ledger/create-token-modal.tsx [hunk 6]
```diff
@@ -148,22 +144,26 @@ export function CreateTokenModal({
...
-              <Input fontFamily="mono" isDisabled value={address} />
+            <FormControl isInvalid={!!errors.destination}>
+              <FormLabel htmlFor="destination">Destination Address</FormLabel>
+              <Controller
+                name="destination"
+                control={control}
+                rules={{
+                  required: true,
+                  validate: {
+                    isManyAddress: (v) => new RegExp(/^m\w{24,}$/).test(v),
```
Comment (stanleyjones) [inline src/pages/services/ledger/create-token-modal.tsx:155]: A more accurate regex for detecting a valid Many address is welcome if we have one (I looked in the spec and `many-rs`).

**Review by stanleyjones — COMMENTED** [review]
Some hopefully useful comments.

**Comment by stanleyjones** [comment]
@fmorency Okay, this one is ready now.

**Review by fmorency — CHANGES_REQUESTED** [review]
I still get an error when trying to create a token using the pre-populated destination field.

Copying and pasting the (same) address works.

**Comment by stanleyjones** [comment]
Okay, that's weird. I'll take a look.

**Comment by stanleyjones** [comment]
Okay, fixed. Re-requesting reviews.

**Comment by stanleyjones** [comment]
This is the CI error:
```
npm ERR! code ECONNRESET
npm ERR! errno ECONNRESET
npm ERR! network Invalid response body while trying to fetch https://registry.npmjs.org/jest-haste-map: aborted
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network 
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/circleci/.npm/_logs/2023-09-07T17_24_52_609Z-debug-0.log

Exited with code exit status 1
```
Going to rerun.

**Review by fmorency — APPROVED** [review]

**Pull request merged**